File: web/calc_http_requester.py
import logging
import requests

DEFAULT_URL = 'http://127.0.0.1:8000'

class CalcHttpRequester:
    def __init__(self, base_url=None, logger=None):
        if base_url is None:
            self.base_url = DEFAULT_URL
        else:
            self.base_url = base_url

        if logger is None:
            logger_name = self.__class__.__name__
            self.logger = make_logger(name=logger_name, file=False, console=False, level=logging.WARNING)
        else:
            self.logger = logger

        self.data = {}

    def get(self, endpoint, params=None):
        if params is None:
            params = {}
        try:
            response = requests.get(
                    f"{self.base_url}/calc/{endpoint}",
                    params=params,
                    timeout=5,
                )
        except requests.exceptions.ConnectionError as exc:
            self.logger.error("Connection to '%s' failed: %s", self.base_url, exc, exc_info=False)
            raise ConnectionError("Connection failed: run api_service")

        if response.status_code != 200:
            self.logger.error("Error while getting sate. Status code: %s. Details: %s", response.status_code, response.text)
            raise requests.HTTPError(f"Bad status code {response.status_code}: {response.json()}")

        return response.json()

    def put_state(self):
        response = requests.put(
                f"{self.base_url}/calc/put_state",
                json=self.data,
                timeout=5,
            )

        if response.status_code == 400 and "detail" in response.json() and "Error" in response.json()["detail"]:
            return response.json()["detail"]['Error']

        if response.status_code != 200:
            self.logger.error("Error while putting sate. Status code: %s. Details: %s", response.status_code, response.text)
            raise requests.HTTPError(f"Bad status code {response.status_code}: {response.json()}")


        data = response.json()
        self.data.update(data)
        self.logger.debug("CalcClient updated data %s", self.data)
        return ""

    def post(self, endpoint, json):
        response = requests.post(
                f"{self.base_url}/calc/{endpoint}",
                json=json,
                timeout=5,
            )

        if response.status_code == 400 and "detail" in response.json() and "Error" in response.json()["detail"]:
            return response.json()["detail"]['Error']

        if response.status_code != 200:
            self.logger.error("Error while making POST request. Status code: %s. Details: %s", response.status_code, response.text)
            raise requests.HTTPError(f"Bad status code {response.status_code}: {response.json()}")


        data = response.json()
        self.data.update(data)
        self.logger.debug("CalcClient updated data %s", self.data)
        return ""

    def delete(self, endpoint, params=None):
        if params is None:
            params = {}

        response = requests.delete(
                f"{self.base_url}/calc/{endpoint}",
                timeout=5,
                params=params,
            )

        if response.status_code != 200:
            self.logger.error("Error while deleting variable. Status code: %s. Details: %s", response.status_code, response.text)
            raise requests.HTTPError(f"Bad status code {response.status_code}: {response.json()}")

        data = response.json()
        self.data.update(data)
        self.logger.debug("CalcClient updated data %s", self.data)

    def save_state(self, session_id):
        """ Ещё один специальный post-запрос для сохранения в файл """
        response = requests.post(
                f"{self.base_url}/calc/save_state",
                params={"session_id": session_id},
                timeout=5,
            )

        if response.status_code != 200:
            self.logger.error("Error while saving sate. Status code: %s. Details: %s", response.status_code, response.text)
            raise requests.HTTPError(f"Bad status code {response.status_code}: {response.json()}")

File: cpp_client/src/requests/calc_http_requester.cpp

#include "calc_http_requester.h"

//Вспомогательная функция (используется также в классе наследнике)
QJsonObject shortJson(QJsonObject data)
{
    const QString keys [2] = {"help_text", "explanations"};
    for (const QString& key : keys)
    {
        if (data.contains(key))
            data[key] = "...";
    }
    return data;
}

CalcHttpRequester::CalcHttpRequester(const QString& baseUrl, int timeout, QObject* parent)
    : QObject(parent)
    , m_executor(new HttpRequestExecutor(baseUrl, timeout, this))
{
}

void CalcHttpRequester::update(const QJsonObject& newdata)
{
    for (const auto& key : newdata.keys())
    {
        m_data[key] = newdata[key];
    }
}

QJsonObject CalcHttpRequester::get(const QString& endpoint, const QVariantMap& params)
{
    try
    {
        return m_executor->get(QString("calc/") + endpoint, params);
    }
    catch (const HttpRequestExecutor::Error& err)
    {
        qCritical() << "Error while getting full sate";
        err.log();
        throw;
    }
    catch (const std::exception& err)
    {
        qCritical() << "Error while getting sate";
        throw;
    }
}

QString CalcHttpRequester::put_state()
{
    try
    {
        QJsonObject newdata = m_executor->put("calc/put_state", m_data);
        this->update(newdata);
        qDebug() << "CalcClient updated data:" << shortJson(m_data);
        return "";
    }
    catch (const HttpRequestExecutor::Error& err)
    {
        if (err.details.isObject())
        {
            QJsonObject detailsObj = err.details.toObject();
            if (detailsObj.contains("Error") && detailsObj["Error"].isString())
                return detailsObj["Error"].toString();
        }
        qCritical() << "Error while putting sate";
        err.log();
        throw;
    }
    catch (const std::exception& err)
    {
        qCritical() << "Error while putting sate";
        throw;
    }
}

QString CalcHttpRequester::post(const QString& endpoint, const QJsonObject& body)
{
    try
    {
        QJsonObject newdata = m_executor->post(QString("calc/") + endpoint, body);
        this->update(newdata);
        qDebug() << "CalcClient updated data:" << shortJson(m_data);
        return "";
    }
    catch (const HttpRequestExecutor::Error& err)
    {
        if (err.details.isObject())
        {
            QJsonObject detailsObj = err.details.toObject();
            if (detailsObj.contains("Error") && detailsObj["Error"].isString())
                return detailsObj["Error"].toString();
        }
        qCritical() << "Error while making POST request";
        err.log();
        throw;
    }
    catch (const std::exception& err)
    {
        qCritical() << "Error while making POST request";
        throw;
    }
}

void CalcHttpRequester::del(const QString& endpoint, const QVariantMap& params)
{
    try
    {
        QJsonObject newdata = m_executor->del(QString("calc/") + endpoint, params);
        this->update(newdata);
        qDebug() << "CalcClient updated data:" << shortJson(m_data);
    }
    catch (const HttpRequestExecutor::Error& err)
    {
        qCritical() << "Error while making DELETE request";
        err.log();
        throw;
    }
    catch (const std::exception& err)
    {
        qCritical() << "Error while making DELETE request";
        throw;
    }
}

void CalcHttpRequester::saveState(int sessionId)
{
    try
    {
        QJsonObject newdata = m_executor->post("calc/save_state", {{"session_id", sessionId}});
    }
    catch (const HttpRequestExecutor::Error& err)
    {
        qCritical() << "Error while saving state";
        err.log();
        throw;
    }
    catch (const std::exception& err)
    {
        qCritical() << "Error while saving state";
        throw;
    }
}
